### ISSUE TITLE
relic_bn: bn_rand_mod() distribution out of bounds

### ISSUE THREAD

**Issue by 1gn1t10n** [relic-toolkit/relic#10]
#### Context
RELIC version: 1d085285a37f75bdbddf1c1c7dce7af3d0bd4d3e
Configure parameters: irrelevant

#### Issue
[`bn_rand_mod(p)`](https://github.com/relic-toolkit/relic/blob/master/src/bn/relic_bn_util.c#L196) is expected to have a uniform distribution in `[0, p)`.
However, assuming `bn_rand(b bits)` has a uniform distribution in `[0, 2^b)`, the output of the current `bn_rand_mod` is uniform in `[1, p]`.
This makes the distribution non-uniform in `[0, p)` as zero is never generated. More importantly, `p` can be generated thus breaking the promise of "modulo p".

#### Possible fix:
Replace the stopping condition with
``` C
void bn_rand_mod(bn_t a, bn_t b) {
        do {
                bn_rand(a, bn_sign(b), bn_bits(b));
        } while (bn_cmp_abs(a, b) != CMP_LT);
}
```

**Comment by dfaranha** [comment]
This function is currently being used for generating private keys in Z_p*, so the correct interval is actually [1, p-1]. Is the documentation clear enough?

**Comment by 1gn1t10n** [comment]
No, you're right about the non-zero condition, I missed it.

There remains the problemthat p might be generated. However I see it has been corrected in commit 9a533ca70f2970fca2345aac2d7eba260912e96b . Thanks!
